### PR TITLE
Classifier: Refactor workflow strings as a map

### LINKED ISSUE #3266: Classifier: background refresh from Panoptes doesn't run when workflow version changes
## Package
lib-classifier

## Describe the bug
When I edit a workflow in the project builder, open tabs that show that workflow don't automatically update to show the new version. The workaround is to reload any open tabs after editing a workflow.

## To Reproduce
1. Open a workflow in the project builder and also in either the dev classifier or the project app, in another tab.
1. Make a change to the workflow and save.
1. Switch back to the open tab with the classifier to check your edits.

## Expected behavior
When you switch back to the classifier, it should update automatically. See the video in this comment: https://github.com/zooniverse/front-end-monorepo/pull/2677#issuecomment-1020207982

## Additional context
There's an `_onWorkflowChange` listener which is supposed to react to any workflow changes. It works when the tests run in Node, but doesn't react to workflow version changes in the browser.
https://github.com/zooniverse/front-end-monorepo/pull/3219#issuecomment-1148390459

### PR DESCRIPTION
**Pull request by eatyourgreens**
`types.frozen` is assumed to be immutable, but `workflow.strings` changes when the locale changes. This refactors `workflow.strings` as an observable map, applying new snapshots of the language strings when the locale changes.

`WorkflowStepStore` is also updated to explicitly update task strings whenever the workflow version changes, so that workflow edits in the project builder are shown in the classifier's task panel.

- Closes #3266.
- Based on #3220, so take a look at that first.

## Package
lib-classifier

## How to Review
Open a workflow in either the dev classifier or the project app. Change the workflow's language strings by either editing the workflow in the project builder or by switching to another language on the project's Classify page.

This PR should also fix background refresh in the classifier, so that the displayed workflow is always the most up-to-date version.

https://user-images.githubusercontent.com/59547/172350935-42703218-c32a-4859-a8be-e489e37f8450.mov

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [x] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [x] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [x] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [x] Can submit a classification
- [x] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Refactoring
- [x] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Playing around with this in the dev classifier, when you change language, `store.workflowSteps.active` is reset to `undefined`. That's unexpected. Also not caught by any tests. Looks like the classification is still there, in the store, though.

**Comment by eatyourgreens** [comment]
One bug might be that the `onURLChange` effect is running whenever `workflowSnapshot` changes, which is going to happen whenever new strings are loaded from `/translations`. I'll look into this some more tomorrow, but that effect should really only depend on the page URL parameters and user roles (for workflows that aren't live.)

**Comment by eatyourgreens** [comment]
@goplayoutside3 I can get this branch to work if I use [`reaction`](https://mobx.js.org/reactions.html#reaction) instead of [`autorun`](https://mobx.js.org/reactions.html#autorun), to detect when the workflow ID or version has changed. `autorun` seems too insensitive, and resets the workflow steps store when any part of the workflow changes, such as `workflow.strings`. What you're seeing is that changing the language runs `_onLocaleChange`, as expected, which then triggers `_onWorkflowChange` to run. `_onWorkflowChange` resets the steps store, and the task panel renders null when there's no active step.

I think this is all a bit confusing to follow, so I'm wondering if there's a better way to do this, without having to rewrite a large portion of the classifier store. `store.workflowSteps` is used throughout the classifier code.

I'll push my changes up so that you can have a look.

**Comment by eatyourgreens** [comment]
Switching out `autorun` for `react` fails 103 tests, so that might not be a great idea.

**Comment by eatyourgreens** [comment]
@goplayoutside3 this should be working now, with `_onWorkflowChange` rebuilding the workflow steps store when `workflow.id` or `workflow.version` changes, but `_onLocaleChange` only rebuilding the task strings when the language changes. There shouldn't be a bug where changing the language completely resets the active step.

**Comment by eatyourgreens** [comment]
I've done some testing of the NextJS app with PH-TESS. When you change locale, with the dropdown menu, the `ClassifierWrapper` component unmounts then mounts again, which then resets the classifier store and deletes any classifications in progress.

**Comment by goplayoutside3** [comment]
> I've done some testing of the NextJS app with PH-TESS. When you change locale, with the dropdown menu, the `ClassifierWrapper` component unmounts then mounts again, which then resets the classifier store and deletes any classifications in progress.

Oh interesting! I was just reviewing again and was able to run this branch locally, visit TESS, draw a transit, and then switch to French (with LocaleSwitcher) without losing my progress 🤔 Could there be something different in our setups?

**Comment by eatyourgreens** [comment]
I forgot that the classifier store is a global variable in the window, so I think that might save us here. When the classifier mounts, it connects to the global `store` variable, if one already exists. I think all workflows should be okay as long as this effect doesn't run and reset the subject queue:
https://github.com/zooniverse/front-end-monorepo/blob/6b26a1850b2f812a7ad83590659fc6c8937c9010/packages/lib-classifier/src/components/Classifier/ClassifierContainer.js#L80-L95

PH-TESS doesn't use session storage, so we don't have to worry about that effect running and resetting the subject queue.

I've got a test workflow here, where I'm seeing the same thing that you see. If I select an answer, then change language, the answer stays selected, even though I can see `ClassifyPage` unmounting and mounting.
https://local.zooniverse.org:3000/projects/es/eatyourgreens/-project-testing-ground/classify/workflow/1483?env=staging

TLDR: I think we're ok as long as `classifierStore` is a global variable, that's only destroyed if you actually reload the browser window.

**Comment by eatyourgreens** [comment]
I think that `useEffect` hook could be improved too, so that the subject queue only gets reset if we're reading the store from local storage ie. after reloading the window. If we're just navigating between different classify pages, then we probably don't want to load a new subject when the `locale` parameter changes in the URL. That sounds like a separate PR, though.

**Comment by eatyourgreens** [comment]
I've opened #3273 to rewrite the `ClassifierContainer` hooks, so that you can change language without resetting the subject queue, but reloading the page will give you a new subject. A side effect of that is that you can navigate to Home or About, navigate back and continue from where you left off. That might be useful on projects with very involved workflows, like transcription projects.

**Comment by goplayoutside3** [comment]
> I've opened #3273 to rewrite the `ClassifierContainer` hooks, so that you can change language without resetting the subject queue, but reloading the page will give you a new subject. A side effect of that is that you can navigate to Home or About, navigate back and continue from where you left off. That might be useful on projects with very involved workflows, like transcription projects.

Sounds good! Does it make most sense to continue review and merge this PR before reviewing #3273?

**Comment by eatyourgreens** [comment]
Yeah. #3237 is a nice-to-have. I think it fixes a bug that only affects my staging project at the moment.

File: packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js [hunk 1]
```diff
@@ -84,15 +85,17 @@ const WorkflowStepStore = types
...
+      if (self.workflow?.id && self.workflow?.version) {
         self.reset()
         self.setStepsAndTasks()
       }
     }
 
     function _onLocaleChange() {
-      if (self.locale && self.workflow?.strings) {
-        self.setTaskStrings()
+      // Update task language strings when the workflow language changes.
```
Comment (goplayoutside3) [inline packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js:96]:
```suggestion
      // Update task translation strings when the classifier's locale changes.
```

File: packages/lib-classifier/src/components/Classifier/Classifier.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import { Provider } from 'mobx-react'
```
Comment (goplayoutside3) [inline packages/lib-classifier/src/components/Classifier/Classifier.js:1]:
```suggestion
```

File: packages/lib-classifier/src/components/Classifier/Classifier.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
...
-import { getSnapshot } from 'mobx-state-tree'
+import { applySnapshot, getSnapshot } from 'mobx-state-tree'
 import PropTypes from 'prop-types'
 import React, { useEffect } from 'react'
 import zooTheme from '@zooniverse/grommet-theme'
```
Comment (goplayoutside3) [inline packages/lib-classifier/src/components/Classifier/Classifier.js:5]:
```suggestion
```

File: packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js [hunk 4]
```diff
@@ -162,20 +166,19 @@ const WorkflowStepStore = types
...
+          const taskStringsSnapshot = {}
+          taskStrings.forEach(([key, value]) => {
+            const newKey = key.slice(prefix.length)
+            taskStringsSnapshot[newKey] = value
           })
           try {
-            applySnapshot(task.strings, strings)
+            applySnapshot(task.strings, taskStringsSnapshot)
           } catch (error) {
             console.error(`${taskKey} ${task.type}: could not apply language strings`)
```
Comment (goplayoutside3) [inline packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js:183]:
```suggestion
            console.error(`${task.taskKey} ${task.type}: could not apply language strings`)
```

File: packages/lib-classifier/src/components/Classifier/Classifier.js [hunk 5]
```diff
@@ -55,20 +56,21 @@ export default function Classifier({
 
   useEffect(function onURLChange() {
     const { workflows } = classifierStore
-    if (workflowID && workflowSnapshot) {
+    if (workflowID) {
       console.log('starting new subject queue', { workflowID, subjectSetID, subjectID })
       workflows.setResources([workflowSnapshot])
```
Comment (goplayoutside3) [inline packages/lib-classifier/src/components/Classifier/Classifier.js:61]: Is there any scenario where `onURLChange()` will be triggered, but `worfklowSnapshot` is undefined or null? Does the WorkflowStore care if the snapshot happens to be undefined or null?

**Review by goplayoutside3 — COMMENTED** [review]
Can confirm that this PR solves #3266. When editing a project’s workflow such as FEM Bug Squashing in the [project builder](https://www.zooniverse.org/lab/16860), viewing the changes [locally](https://local.zooniverse.org:3000/projects/blicksam/fem-bug-squashing/classify/workflow/19629?env=production) causes the classifier to re-render on focus and resets progress in the WorkflowStepStore.
 

Is preserving task progress during locale switch intended as part of this PR? 

- I can confirm that PH-TESS and i-fancy-cats progress is preserved when using the LocaleSwitcher (between English and Test Language).
- Projects that use subject set selection in a workflow do not preserve progress. Please check out HMS NHS where text typed into the TaskArea is lost on locale change, and Poets & Lovers where transcription lines are lost on locale change.
- There's a quirk when viewing Beyond Borders where progress is preserved, but the last edited transcription modal opens halfway off screen when locale is changed.

File: packages/lib-classifier/src/components/Classifier/Classifier.js (hunk 5 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/Classifier.js:59]: Sure. The snapshot will be undefined while we’re waiting for a workflow to load, but I think the container component waits until the workflow has loaded before mounting the classifier. 

The tests don’t include the container at the moment. The mocks in the `Classifier` should match what we expect the container to pass down but it would probably be more robust to include `ClassifierContainer` in the tests, and test against a mock Panoptes API.

File: packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStepStore/WorkflowStepStore.js:96]: This function reacts to `self.workflow.strings`. If it reacts to the store locale, then it fires before the workflow strings change. Now that `workflow.strings` is observable, we can observe it directly for changes. 

Maybe I should rename it?

**Comment by eatyourgreens** [comment]
Beyond Borders sounds like https://github.com/zooniverse/front-end-monorepo/issues/2272. You can probably also trigger that on a drawing task by going forward to the next step, then back to the drawing step. I don’t know if there’s an issue already for the popup opening in the wrong place. 

I’ll take a look at the text task. I can’t think of any reason why it would be deleted from the store.

File: packages/lib-classifier/src/components/Classifier/Classifier.js (hunk 5 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/Classifier.js:59]: Without looking at the code for the store model, I’d guess that passing null here would throw an error. MobX State Tree is strict about [types](https://mobx-state-tree.js.org/overview/types) and I don’t know if the classifier models allow null or undefined for Panoptes resources, like workflows.

**Comment by eatyourgreens** [comment]
HMS NHS and Poets & Lovers both use prioritised workflows, so I can see in the dev console that they're triggering this side effect, which resets the subject queue when the classifier mounts:
https://github.com/zooniverse/front-end-monorepo/blob/6b26a1850b2f812a7ad83590659fc6c8937c9010/packages/lib-classifier/src/components/Classifier/ClassifierContainer.js#L80-L95

I've removed that effect in #3237.

File: packages/lib-classifier/src/components/Classifier/Classifier.js (hunk 5 above)
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/Classifier.js:59]: Sounds good, thank you for the clarification 👍

**Comment by goplayoutside3** [comment]
> That should only run when you reload the page. I've removed that effect in #3237, so that should fix those workflows.

#3237 is a dependency bump - did you mean #3273 about the ClassifierContainer hooks? And good to know! I think this PR is ready to merge then.

**Review by goplayoutside3 — APPROVED** [review]

**Comment by eatyourgreens** [comment]
Yes! #3273. I just added a couple of tests to that PR. I'm not sure if any of those transcription projects are planning to use translations, so I'm not sure if fixing them is high priority.
